lib/recursive-methods.rb

```diff
@@ -1,49 +1,87 @@
 # Authoring recursive algorithms. Add comments including time and space complexity for each method.
 
-# Time complexity: ?
-# Space complexity: ?
+# Time complexity: O(n)
+# Space complexity: O(n)
 def factorial(n)
-    raise NotImplementedError, "Method not implemented"
+  raise ArgumentError if n < 0
+  return 1 if n == 1 || n == 0
+  return n * factorial(n - 1)
 end
 
-# Time complexity: ?
-# Space complexity: ?
+# Time complexity: O(n)
+# Space complexity: O(n)
 def reverse(s)
-    raise NotImplementedError, "Method not implemented"
+  return s if s.length <= 1
+  reversed_str = reverse(s[1..-1])
+  reversed_str += s[0]
+  reversed_str
 end
 
-# Time complexity: ?
-# Space complexity: ?
+# Time complexity: O(n)
+# Space complexity: O(1)
 def reverse_inplace(s)
-    raise NotImplementedError, "Method not implemented"
+  return s if s.length < 1
+  return (s[-1] + reverse_inplace(s[0...-1]))
 end
 
-# Time complexity: ?
-# Space complexity: ?
+# Time complexity: O(n)
+# Space complexity: O(n)
 def bunny(n)
-    raise NotImplementedError, "Method not implemented"
+  if n == 0
+    return 0
+  elsif n == 1
+    return 2
+  end
+  return 2 + bunny(n - 1)
 end
 
-# Time complexity: ?
-# Space complexity: ?
+# Time complexity: O(n)
+# Space complexity: O(1)
 def nested(s)
-    raise NotImplementedError, "Method not implemented"
+  if s == ""
+    return true
+  elsif s.length == 1
+    return false
+  elsif s[0] == s[-1]
+    return false
+  else
+    return nested(s[1..-2])
+  end
 end
 
-# Time complexity: ?
-# Space complexity: ?
+# Time complexity: O(n)
+# Space complexity: O(1)
 def search(array, value)
-    raise NotImplementedError, "Method not implemented"
+  if array.empty?
+    return false
+  elsif array[0] == value
+    return true
+  else
+    return search(array[1..-1], value)
+  end
 end
 
-# Time complexity: ?
-# Space complexity: ?
+# Time complexity: O(n)
```

Comment:
O(n^2) for both time and space complexity.  You could do it with O(n) using a helper method and current left_index and right_index parameters.